### DISCUSSION TITLE
Connecting the HTML frontend to FastAPI

### DISCUSSION THREAD

**Discussion by Saviour1001** [IsaiahT-Tech/fastapi-series#6] [Q&A] [answered]
![git](https://user-images.githubusercontent.com/71517788/107735111-f5802b80-6d24-11eb-89e3-e5b3f868c131.PNG)

This is the requests I am putting from the python file ... the requests are working well from swagger but I am still not able to connect the HTML with it. Below it has the HTML file in the folder Templates..Could someone tell me what should I do to solve the query. 
The github repo: https://github.com/Saviour1001/practicehtml/tree/master/Game%20on

**Comment by Saviour1001** [comment, marked as answer]
Got the fetch API working but when I am sending request from HTML to fastapi , it shows the error 422 unprocessable entity ... the fetch API is sending a json response ...

**Reply by IsaiahT-Tech** [reply]
Looking at your code, it seems like you are using JavaScript, are you fine with using Jinja2 or should it be JavaScript?

**Reply by Saviour1001** [reply]
There are no compulsions .... could use anything .. i was familiar more with javascript .. thus used that ... but apparently didn't got it working

**Reply by IsaiahT-Tech** [reply]
Hi, I have made a sample of code that may help you with connecting your HTML to your backend FastAPI. This is using Jinja2 Templates / HTML for the frontend. This is an example, I hope this helps. 

`main.py`

``` python
from typing import List

from fastapi import FastAPI, Depends, Request, Form
from pydantic import BaseModel, Field
import databases
import sqlalchemy
from fastapi.templating import Jinja2Templates
from fastapi.responses import HTMLResponse

DATABASE_URL = "sqlite:///./store.db"

metadata = sqlalchemy.MetaData()

database = databases.Database(DATABASE_URL)

register = sqlalchemy.Table(
    "register",
    metadata,
    sqlalchemy.Column("id", sqlalchemy.Integer, primary_key=True),
    sqlalchemy.Column("name", sqlalchemy.String(500)),
)

engine = sqlalchemy.create_engine(
    DATABASE_URL, connect_args={"check_same_thread": False}
)

metadata.create_all(engine)

templates = Jinja2Templates(directory="templates")

app = FastAPI()

@app.on_event("startup")
async def connect():
    await database.connect()

@app.on_event("shutdown")
async def shutdown():
    await database.disconnect()

@app.get('/')
async def read_root(request: Request):
    query = register.select()
    rows = await database.fetch_all(query)
    return templates.TemplateResponse("index.html", {"request": request, "rows": rows}) 

@app.post('/')
async def create(request: Request, name: str = Form(...)):
    query = register.insert().values(name=name) 
    await database.execute(query)
    query = register.select()
    rows = await database.fetch_all(query)
    return templates.TemplateResponse("index.html", {"request": request, "rows": rows})
```

`templates/index.html`

```html

<html> 
    <head>
        <title>Test</title>
    </head>
    <body>
        <form method="POST" action="/" autocomplete="off" style=text-align:center>
            <input type="text" name="name" required>
            <input type="submit">
        </form>
    
        {% for register in rows %}
            <h2>Title: {{ register.name }} (ID: {{ register.id }})</h2>
        {% endfor %}
    </body> 
</html>


```

**Reply by Saviour1001** [reply]
Thanks a lot ..... my code just was using Depends as a response model which was leading to the error of 422 unprocessable entity as I was sending a JSON ... removed the Depends and yeah it got working ! Thankyou for the help

**Reply by IsaiahT-Tech** [reply]
Your welcome. Thank you! Yes, I encountered that error too with Depends, and it had to be removed. Thank you!
